### PR TITLE
Fix "Open next file"

### PR DESCRIPTION
**Pull request by hius07**
Finished books are always skipped while searching the next file.

Closes https://github.com/koreader/koreader/issues/11266. Closes https://github.com/koreader/koreader/issues/11267.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11272)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/b949d07f4fbf7ffb0a145684f98f196bb0492aea/frontend/apps/reader/modules/readerstatus.lua#L126-L130
Since we do not destroy the document anymore, I am not sure if we need a delay. `switchDocument` would take care of this. @NiLuJe ?

File: frontend/apps/reader/modules/readerstatus.lua [hunk 1]
```diff
@@ -169,12 +169,8 @@ function ReaderStatus:openFileBrowser()
...
 
 function ReaderStatus:onOpenNextDocumentInFolder()
-    local FileManager = require("apps/filemanager/filemanager")
-    if not FileManager.instance then
-        self.ui:showFileManager()
-    end
-    local next_file = FileManager.instance.file_chooser:getNextFile(self.document.file)
-    FileManager.instance:onClose()
+    local FileChooser = require("ui/widget/filechooser")
+    local next_file = FileChooser:getNextFile(self.document.file)
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readerstatus.lua:173]: Err, FileChooser isn't a singleton, we *do* need an instance of it for this to work?

**Comment by NiLuJe** [comment]
> Since we do not destroy the document anymore, I am not sure if we need a delay. `switchDocument` would take care of this. @NiLuJe ?

Err, why wouldn't we destroy the document anymore? (Because that sounds like a very big problem to me, unless I'm missing something ;p).

File: frontend/apps/reader/modules/readerstatus.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua:173]: And if that actually worked, it means we have a much bigger problem of something storing stuff in the class's `item_table`... *somewhere* ?

Comment (hius07) [reply frontend/apps/reader/modules/readerstatus.lua:173]: It's not a class's item_table, it's a local array that is destroyed after function is finished.

**Comment by hius07** [comment]
Previously we called `showFileManager`, to find the next file, no we don't.

**Comment by hius07** [comment]
Should `switchDocument` be always called within `nextTick`?

File: frontend/apps/reader/modules/readerstatus.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua:173]: Err, I'm definitely seeing that access `self.item_table` as populated in `init`, unless I'm missing something obvious?


https://github.com/koreader/koreader/blob/b949d07f4fbf7ffb0a145684f98f196bb0492aea/frontend/ui/widget/filechooser.lua#L233-L237

=>

https://github.com/koreader/koreader/blob/b949d07f4fbf7ffb0a145684f98f196bb0492aea/frontend/ui/widget/filechooser.lua#L561-L574

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua:173]: Which is precisely why this was spinning up `FileManager` before ;).

Comment (hius07) [reply frontend/apps/reader/modules/readerstatus.lua:173]: Please look at the diffs.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua:173]: Oh, see? Something obvious ;D.

My bad, not at the computer, so I completely missed the function name :D.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua:173]: And `show_finished` is a class member and used as such, so this is perfectly fine, cool! :}

**Comment by NiLuJe** [comment]
> Should `switchDocument` be always called within `nextTick`?

Unless I missed another obvious bit of diff context, it depends on the context ;).

The bit you quoted in https://github.com/koreader/koreader/pull/11272#issuecomment-1869541991 comes from an Event handler (onEndOfBook), and the concern was that it may be *followed* by other events that *need* a Document instance, so anything that would kill it need to be delayed to let those events comes through unscathed ;).

So, TL;DR: if within an event handler, yes; otherwise, no.

File: frontend/apps/reader/modules/readerstatus.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua:173]: And this is also the only caller of `getNextFile`, so we don't have to worry about other call sites doing it differently.

File: frontend/ui/widget/filechooser.lua [hunk 2]
```diff
@@ -559,14 +559,20 @@ end
 
 -- Used in ReaderStatus:onOpenNextDocumentInFolder().
 function FileChooser:getNextFile(curr_file)
+    local show_finished = self.show_finished
```
Comment (NiLuJe) [inline frontend/ui/widget/filechooser.lua:562]: I wouldn't mind making it explicit that this is a class member by *always* accessing it as `FileChooser.` (ditto for `show_unsupported`,  e.g., in `FileChooser:show_file`)

File: frontend/ui/widget/filechooser.lua [hunk 3]
```diff
@@ -559,14 +559,20 @@ end
 
 -- Used in ReaderStatus:onOpenNextDocumentInFolder().
 function FileChooser:getNextFile(curr_file)
+    local show_finished = FileChooser.show_finished
+    FileChooser.show_finished = true
```
Comment (NiLuJe) [inline frontend/ui/widget/filechooser.lua:563]: Shouldn't this be *false* if we want to always ignore finished books here?

Comment (NiLuJe) [reply frontend/ui/widget/filechooser.lua:563]: Oh, okay, I get it.

We need it set to actually collect finished books, but the skipping is handled manually later, instead of relying on show_file ;).

Comment (hius07) [reply frontend/ui/widget/filechooser.lua:563]: Exactly. The user just has marked the book as finished, but we must keep it in the item_table to find the next one in accordance with the sorting order.
